utils/src/tree/GenericParentTreeItem.ts

```diff
@@ -0,0 +1,54 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { TreeItemCollapsibleState } from "vscode";
+import * as types from '../../index';
+import { AzExtParentTreeItem } from "./AzExtParentTreeItem";
+import { AzExtTreeItem } from "./AzExtTreeItem";
+import { IAzExtParentTreeItemInternal } from "./InternalInterfaces";
+
+interface GenericParentTreeItemOptions {
+    childTypeLabel?: string;
+    contextValue: string;
+    iconPath?: types.TreeItemIconPath;
+    initialCollapsibleState?: TreeItemCollapsibleState;
+    label: string;
+    suppressMaskLabel?: boolean;
+
+    compareChildrenImpl?(item1: AzExtTreeItem, item2: AzExtTreeItem): number;
+    loadMoreChildrenImpl?(clearCache: boolean, context: types.IActionContext): Promise<AzExtTreeItem[]>;
+}
+
+export class GenericParentTreeItem extends AzExtParentTreeItem implements types.GenericParentTreeItem {
+    public childTypeLabel?: string;
+    public contextValue: string;
+    public label: string;
+    public suppressMaskLabel?: boolean;
+
+    public readonly initialCollapsibleState: TreeItemCollapsibleState;
+
+    constructor(parent: IAzExtParentTreeItemInternal | undefined, readonly options: GenericParentTreeItemOptions) {
```

Comment:
Question: Should we allow parent to be passed as `undefined`?

There is one known issue with calling `this.treeDataProvider` if passed as undefined as `treeDataProvider` is not set and `nonNullProp` will throw an error.

![image](https://github.com/microsoft/vscode-azuretools/assets/40250218/0f90a1d5-c36e-479e-9331-396773c6f3e5)

![image](https://github.com/microsoft/vscode-azuretools/assets/40250218/463641bb-a5d6-429f-be1f-c2b60110f719)

Reply:
Discussed in engineering office hours